### DISCUSSION TITLE
threadstatic keyword

### DISCUSSION THREAD

**Discussion by znakeeye** [dotnet/csharplang#6398] [Language Ideas]
A less known feature in `.NET` is the [ThreadStaticAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.threadstaticattribute?view=net-6.0). When used correctly, it allows for significant performance improvements. Though, for semi-experienced developers, I believe the syntax is slightly too complicated for this feature to get the attention it deserves. Thus, my proposal below.

**NEW: Using threadstatic**
```c#
private static class Helper<TKey, TDelegate>
    where TKey : notnull
    where TDelegate : Delegate
{
    private threadstatic Dictionary<TKey, TDelegate> Cache { get; } = new();
}
```

**OLD: Using ThreadStaticAttribute**
```c#
private static class Helper<TKey, TDelegate>
    where TKey : notnull
    where TDelegate : Delegate
{
    [ThreadStatic]
    private static Dictionary<TKey, TDelegate>? cache;
    
    internal static Dictionary<TKey, TDelegate> Cache => cache ??= new();
}
```

Note that you have to remember _not_ to instantiate the backing field. It's easy to forget this, which is why you have this important note in the [Remarks](https://docs.microsoft.com/en-us/dotnet/api/system.threadstaticattribute?view=net-6.0#remarks) section:

> Do not specify initial values for fields marked with ThreadStaticAttribute, because such initialization occurs only once, when the class constructor executes, and therefore affects only one thread. If you do not specify an initial value, you can rely on the field being initialized to its default value if it is a value type, or to null if it is a reference type.

**Key benefits**
- Improved syntax, less typing.
- Developers will learn to use `threadstatic` - where appropriate - instead of `ConcurrentDictionary`, `lock()` and other less performant threading mechanisms.
- Visual Studio could hint about this. With some caveats of course. E.g. _"Consider using threadstatic if this class is accessed by a limited number of threads..."_
- Easy to implement. Just some syntactical sugar around the `ThreadStaticAttribute`.

**Comment by HaloFour** [comment]
I don't see why `ThreadStaticAttribute` is too complicated here.  It's exactly two additional characters, so verbosity is definitely not an issue.

The use of attributes to denote language/runtime features without the need of an additional keyword is very normal, it's one of the reasons that attributes exist in the language to begin with and this has been the state for 20+ years.  Semi-experienced developers of the C# language would already need to know this in order to use many other features of the language effectively.

This is also not the only attribute to effect language/runtime behavior in this manner.  There are quite a few others, many which have also been there since C# 1.0, like `SerializedAttribute`, `DllImportAttribute` or `StructLayoutAttribute`.  Are you suggesting that the language team should reverse their position on the use of attributes here and instead design explicit language syntax for all features supported by those attributes?

**Reply by znakeeye** [reply]
I see your points. From a technical perspective, you are indeed right. From a maybe-irrelevant UX perspective, I believe my argument holds true.

**Comment by svick** [comment]
I think `[ThreadStatic]` is a niche feature that does not deserve its own keyword and the benefits of making the syntax more succinct are tiny.

On the other hand, ensuring `[ThreadStatic]` is used correctly is worthwhile, especially since it can be done much more easily using analyzers.

Note that there should be a message if you're incorrectly initializing a `[ThreadStatic]` field, see https://github.com/dotnet/roslyn-analyzers/pull/5920. Though that message does not show up for me and I'm not sure why.

**Reply by znakeeye** [reply]
I guess improved analyzers is the way to go then.

**Comment by iam3yal** [comment]
Most developer that needs it knows it exists, we really don't need a keyword for this but more importantly a keyword wouldn't make it more visible.

Generally you don't spend time to learn things just because... you spend the time to look for things when you need them and imo the attribute gets the job done.

**Comment by qrli** [comment]
Thread-static is mostly the wrong choice since we got async/tasks, where it does not really work correctly. Instead, `AsyncLocal<T>` should be the normal choice.
Thread-static is now only useful in some very low-level code, or very old legacy code which does not use any async or tasks.

**Comment by huoyaoyuan** [comment]
Adding keyword makes additional -100 points onto the initial minus points of modifying language.
Keywords make a feeling that some compile time check is accomplished. So it's natural to make `required` a keyword.

**Comment by acaly** [comment]
Maybe this is a good place for a source generator, if you really think it helps (obviously many people here don't think so).

You write 

```c#
[GenerateThreadStatic]
public partial Dictionary<string, object> Cache { get; }
```

Generator writes:
```c#
[ThreadStatic]
private Dictionary<string, object>? _cache;
partial Dictionary<string, object> Cache => _cache ??= new();
```

This requires the partial property feature #3412.

**Reply by svick** [reply]
Or, with semi-auto properties, which is already being implemented and is currently planned for C# 12, I think you will be able to write:

```c#
[field: ThreadStatic]
static Dictionary<string, object> Cache => field ??= new();
````

**Reply by znakeeye** [reply]
> Or, with semi-auto properties, which is already being implemented and is currently planned for C# 12, I think you will be able to write:
> 
> ```cs
> [field: ThreadStatic]
> static Dictionary<string, object> Cache => field ??= new();
> ```

Didn't know about that field attribute thing. Pretty cool!
